packages/doenetml-worker-rust/lib-doenetml-core/src/core/general_prop/math_prop.rs

```diff
@@ -1 +1,385 @@
+/// in the expression rather than the actual math values.
+/// The key is that the expression template doesn't change if the math values change,
+/// eliminating the need for the relatively expensive re-parsing of the string.
+///
+/// Parameters:
+/// - `maths_and_strings`: the math and string values forming the expression
+/// - `parser`: an enum specifying whether to use the latex or text parser to create the `MathExpr`
+/// - `split_symbols`: if true, the parse will split multi-character variables that don't contain digits
+///   into the product of their characters
+/// - `function_symbols`: the list of variable names that will be treated as a function if they are followed
+///   by parentheses.
+///
+/// Return a tuple of:
+/// - the expression template
+/// - the generated codes for each math value
+fn calc_expression_template(
+    maths_and_strings: &mut [MathOrString],
+    parser: MathParser,
+    split_symbols: bool,
+    function_symbols: &[String],
+) -> (MathExpr, Vec<String>) {
+    let string_children = Vec::from_iter(maths_and_strings.iter().filter_map(|prop| match prop {
+        MathOrString::String(string_prop) => Some(string_prop),
+        MathOrString::Math(_) => None,
+    }));
+
+    // let code_prefix be a sequence of "m"s long enough so that
+    // it does not appear in any string props
+    let code_prefix = {
+        let mut code_prefix = String::from('m');
+        string_children.iter().for_each(|str_prop| {
+            while str_prop.get().contains(&code_prefix) {
+                code_prefix.push('m');
+            }
+        });
+        code_prefix
+    };
+
+    // concatenate the maths and strings, with maths replaced by the generated codes
+    let (template_string, math_codes) =
+        create_template_string(maths_and_strings, code_prefix, parser);
+
+    // parse this string into a `MathExpr`, which is the expression template
+    let expression_template = match parser {
+        MathParser::Latex => MathExpr::from_latex(template_string, split_symbols, function_symbols),
+        MathParser::Text => MathExpr::from_text(template_string, split_symbols, function_symbols),
+    };
+
+    (expression_template, math_codes)
+}
+
+/// Concatenate all string and math values, only replacing the math values with generated codes.
+///
+/// The code for each math value is `code_prefix` followed by a number.
+///
+/// Parameters:
+/// - `maths_and_strings`: the math and string values forming the expression
+/// - `code_prefix`: the beginning of each generated code used to represent the math values
+/// - `parser`: an enum specifying whether we will use the latex or text parser to create the `MathExpr`
+///
+/// Returns a tuple of:
+/// - the template string that will be parsed into the expression template
+/// - the math codes used to represent each math value
+fn create_template_string(
+    maths_and_strings: &mut [MathOrString],
```

Comment:
Mutable values for functions with a return value make me unhappy. Why do you need this `mut`?

Reply:
OK, I at least got rid of this `mut` by moving the recording of data viewing outside of the `calculate` function.